rfcs/side-effect-lifecycle-scheduling.md

```diff
@@ -0,0 +1,124 @@
+
+>Draft
+
+# RFC: Encoding Side Effect Lifecycles as Subgraphs
+
+This RFC presents an API for scheduling around the lifecycle of side effects. This assumes that exclusive systems can be labeled and ordered in the same way parallel systems can and does not include how the executor might need to be changed to support this API.
+
+## Motivation
+
+A system may produce a side effect that other systems must read, write, and potentially consume. Systems that do this have a very specific order since the side effect must be produced before it can be acted on and cannot be acted upon after they have been consumed. An example of this would be commands. Commands are created in one system and then consumed later when they're applied to the world. If you tried to read the command queue after the commands have been applied, you would only find an empty queue.
+
+We can express this required ordering with labels, before, and after, but they lack semantic meaning. This is an api for creating subgraphs that express the required ordering between producers, consumers, readers, and writers.
+
+## User Facing Description
+
+### Scheduling Side Effects API
+* `.produces(Label)` System produces a side effect of type `Label`
+* `.consumes(Label)` No read/modifies systems can run after this system. An example that would use this is the system that applies commands to the world.
+* `.reads(Label)` system has read access to side effect and must run after the producers and before the consumer. An example of this type of system would be systems that depend on a run criteria.
```

Comment:
Currently, reads and writes have identical behavior. They should either be merged, or readers should always run after writers (a fairly standard pattern).